### PR TITLE
Frontend tokens

### LINKED ISSUE #1558: Allow configuring an API key to use for server-side Openverse API requests
## Problem

Currently the production frontend exempts itself from the API throttling by auto-magically adding its IP address to the list of exempted IPs in our production Redis. This has several bad problems:

1. It requires managing bespoke throttling classes (see https://github.com/WordPress/openverse-api/issues/585) to enable this behavior
2. It requires bad and leaky permissions to be given to the frontend EC2 instances for it to be able to access and write to the Redis cache that it otherwise does not need access to. This creates a new potential vector for someone to attack our API.

We don't want to replicate the second part in our ECS deployment (indeed, it would be a pain in the butt!) so let's just remove this anti-pattern altogether and use API keys which are expressly designed for this purpose.

## Description

Add a new environment variable to the Nuxt app, `OPENVERSE_API_KEY` that is used for server-side API requests. Ensure that it is _only_ used for server side requests and that the key is under no circumstances leaked to the client.

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes #1558 by @sarayourfriend 

## Description

This PR adds a mechanism to send authenticated requests when an API key has been provided (and unauthenticated ones) when not.

This means that API calls during SSR will be authenticated, and not throttled, but API calls from the client (subsequent calls) will be throttled as an authenticated user because they won't have any API key set.

## Testing Instructions

0. Check out the repo.
1. Follow instructions in `TESTING_GUIDELINES.md` for setting up your local environment with API authentication tokens after registering an app on your local dev server or on staging.
    * Be sure to use the version of the document from this branch as I've introduced new documentation for how to do this!
2. Visit the application as normal and confirm in the logs you can see the API token flow. Especially confirm that the _first_ request does wait for the token to come back before continuing.

Furthermore, test using the docker image:

```
docker build . --build-arg RELEASE=v0-test --build-arg API_URL=https://api-dev.openverse.engineering/ -t openverse-frontend:latest
docker run --env-file .env -p 8443:8443/tcp openverse-frontend:latest
```

> **Note:** Earlier this testing instructions read to use a locally running API with the docker container. This doesn't work due to CORS issues and having different URLs for the client and server to hit the API (due to the different network contexts they're running in). To test docker, please use a live API instance. To test against a local API, run the Nuxt app outside of docker using `pnpm prod` and an `.env` file set up to point to your local API.

Make sure to visit the docker-running service at `localhost:8443` and _not_ the IP address that nuxt spits out. That IP address is internal to the docker network. For some reason it still lets some requests through but most of them will fail (including static assets). If you stick to `localhost:8443` it works fine though.

When making requests to the docker image, you will see more than one API token being generated now in the logs, depending on the number of CPUs docker has access to. That is because pm2 runs a worker per CPU it finds. If you're using docker desktop and it is not configured with multiple CPUs then you will not see this behavior. Please enable at least 2 CPUs so that you can test this!

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by sarayourfriend** [comment]
If we put the client key and secret into the `privateRuntimeConfig`, could we put a function into context instead of the actual API key? Otherwise, we could add a server-only middleware that checks a module-level variable that stores the API key and expiration time.

```ts
// ~/plugins/api-authentication.server.ts

import type { Plugin, Context } from '@nuxt/types'

interface State {
  apiKey: string
  expiration: Date
}

const local: { state: State | null } = {
  state: null,
}

const getApiKey = async (ctx: Context): State => {
  /* ... */
}

const expiresSoon = (s: State): boolean => {
  /* ... */
}

const apiAuthentication: Plugin = async (ctx, inject) => {
  if (!local.state || expiresSoon(local.state)) {
    local.state = await getApiKey(ctx)
  }

  inject('openverseApiKey', local.state.apiKey)
}
```

**Comment by sarayourfriend** [comment]
It just sucks to have to have the `apiService` wired up and dependent on Nuxt state. It's suck a hack, like you said, and it infects _everything_.

**Comment by sarayourfriend** [comment]
With the code above I suggested I think `getApiKey` would need an async lock on it as well to prevent multiple requests per worker to get a token :thinking:

**Comment by dhruvkb** [comment]
@sarayourfriend I tried this with a server-side plugin and it was refreshing the access token on every request. It seems to me that plugins are reinitialised on every request and so their state always starts as a blank one.

**Comment by sarayourfriend** [comment]
> it was refreshing the access token on every request

This is odd. I can reproduce it in the dev build but in the production build it works as expected (like a normal node module resolution).

I'll have some changes up soon.

**Comment by sarayourfriend** [comment]
Changes pushed. They work if you do the following:

1. Create a `.env` file with `API_URL` pointing to your local API
2. Add `API_CLIENT_ID` and `API_CLIENT_SECRET` with client credentials for a local API exempted token
3. Build and run the server: `env $(cat .env) pnpm build && env $(cat .env) pnpm start`

You can try the `/secret-test` page but also just check out the console as I've added additional logging to make it easy to track the request through the plugin.

**Comment by sarayourfriend** [comment]
Unit tests are currently failing due to missing API token so we'll have to sort out a way around that. Pushing the token into context makes the component code that relies on it a lot simpler. No need to do conditional checks or handle asynchrony.

**Comment by sarayourfriend** [comment]
This PR is ready for review now and the description has been updated with up-to-date testing instructions. You can also visit the `/secret-test` page, but I will remove that before merging the PR.

File: src/data/api-service.ts
Comment (obulat) [inline src/data/api-service.ts:52]: Nice!

File: TESTING_GUIDELINES.md
Comment (krysal) [inline TESTING_GUIDELINES.md:35]: Is this necessary? I was using and `.env` file before this PR, and variables were getting in fine.

Comment (sarayourfriend) [reply TESTING_GUIDELINES.md:35]: Oh really? I didn't realize but I guess Nuxt has it built in! Thanks for pointing it out. I will revise the instructions.

Comment (sarayourfriend) [reply TESTING_GUIDELINES.md]: Fixed in https://github.com/WordPress/openverse-frontend/pull/1569/commits/0976ebd6bfa10b7aa937c67307ca300755ed76cf

Thanks again!

**Review by dhruvkb — COMMENTED** [review]
I ran this locally and the "happy flow" seems to be working pretty well. I'm currently trying to understand how we could test these events:
- access tokens expiring
- token fetch failing (perhaps invalid credentials)
- requests being blocked till the token is refreshed

File: src/plugins/api-token.server.ts
Comment (sarayourfriend) [reply src/plugins/api-token.server.ts:87]: It would. Ideally we'd have some way to monitor this. We could dispatch a Sentry error? Alternatively CloudWatch log insights could be used to build an alarm... but given we already have a workflow around Sentry I'll try to add that!

**Comment by sarayourfriend** [comment]
I think we could test all of those using the axios mock adapter. Thanks for pointing them out. I will write those tests.

**Review by krysal — APPROVED** [review]
This works well on my end, even the docker build on mac with some extra packages. I have some questions, but nothing worth blocking, I believe 👏

Perhaps when the token expires we can force a hard refresh at the next request, so the request is done server side, and then the token gets refreshed? It's not ideal but could be a solution in the interim.

File: src/pages/search.vue
Comment (sarayourfriend) [reply src/pages/search.vue:155]: Because you can't return functions in `setup` and also use Options API for some things or else Nuxt will raise the "cannot stringify unbound function" error in the console. See https://github.com/WordPress/openverse-frontend/issues/1237 for where we've fixed this before.

File: nuxt.config.ts
Comment (sarayourfriend) [reply nuxt.config.ts:174]: I don't think this is a valid todo. All of the variables in `env` are required by the client and are also safe to send to the client.

**Comment by sarayourfriend** [comment]
> Perhaps when the token expires we can force a hard refresh at the next request, so the request is done server side, and then the token gets refreshed? It's not ideal but could be a solution in the interim.

I don't I understand the suggestion or what problem it is trying to solve. The token requests always happen on the server side and the token is never sent to the client (this would be unsafe). When the token is about to expire, we pre-emptively request a new one and all other requests that come in will wait until the new token is retrieved.

@krysal I've also dismissed your review because I made some non-trivial changes to the logic and I want to make sure the additional documentation in the code is reviewed and comprehensible. Can you re-review with an eye towards that?

I've also added extensive unit tests, some of which are rather complicated due to juggling asynchrony in the tests themselves. @dhruvkb please review the tests and let me know if they are sufficient.

If anyone can think of additional tests to write, please let me know.

File: src/data/api-service.ts
Comment (sarayourfriend) [reply src/data/api-service.ts:87]: I think `isVersioned` here assumes that the api service being created will only be used for a particular set of resources or requests rather than all requests. Because we create multiple axios clients for different api service usages, we pass `isVersioned` when we want the URLs to automatically include the v1 string.

However, I also disagree with this change. @dhruvkb can you explain the benefit of this implementation over explicitly including the version string at the request site? If we introduce new versions of existing endpoints or have api services used for mixed version requests in the same cycle then this implementation will either have to be reverted back to the explicit version or be significantly increase the complexity of this code.

I prefer explicit request URLs personally, they are easier to search for in the app and make zero assumptions about the future or current versions of API endpoints that will or are being used.

Comment (sarayourfriend) [reply src/data/api-service.ts:87]: Well, I think undoing this refactor is actually more trouble than it is worth and will complicate this PR. The existing "versioned API" approach we took was effectively the same and had the same bad assumptions (that `v1` is the only version and that all requests performed by a given API service instance will all reference the same version of endpoint).

I personally find the current approach with the `getResourceSlug` function over-complicated, sneaky, and bakes assumptions into the base `api-service.ts` module that are not necessarily true.

Either the API service instances are only for making requests to the media endpoints (and we should use a plain axios instance for anything else) or we have to reconsider how the API service is used in the first place.

I think it's potentially an anti-pattern, the way it's currently implemented, and we might want to consider stuffing it into Nuxt context, either by hand-spinning our own module or using the existing nuxt/axios module.

This would have some significant benefits:

1. We could insert the plugin after the API token request plugin so that it can automatically include the access token on the server. This allows downstream requesters to completely ignore the api token detail. Currently all api service instantiators have to consider the api token. This duplicates the work and complicates the downstream code.
2. We can eliminate the complexity in the `api-service` module completely. The only issue that remains is the mapping of the media type to the endpoint names, but this can easily be put into a function or a static mapping for reference by requesters to construct API urls.

I'll open an issue for this. For now I think this should remain as is with @dhruvkb's refactors because it keeps the existing assumptions, but makes them ever so slightly more flexible to accommodate non-versioned API requests.

Comment (sarayourfriend) [reply src/data/api-service.ts:87]: https://github.com/WordPress/openverse-frontend/issues/1591

File: src/data/api-service.ts
Comment (obulat) [inline src/data/api-service.ts:16]: Is this something that needs to be done in the API? If so, maybe we can add a link to the API issue here to follow up?

File: src/pages/secret-test.vue
Comment (obulat) [inline src/pages/secret-test.vue:53]:
```suggestion
      data.apiRes.message = (await apiService.query('rate_limit', {}))
```

Comment (obulat) [reply src/pages/secret-test.vue:53]: Currently, this gives an error because `noTrailingSlash` array contains `rate_limit`, not `rate_limit/` with trailing slash, so `getResourceSlug` not only does not _not add_ a slash, but adds one more.

**Comment by obulat** [comment]
I cannot build docker using the commands in the testing instructions because python cannot be found in the path.

<details>

<summary>Error message: installing deasync requires Python</summary>

```shell
#9 26.19 .../node_modules/@sentry/cli install: Done
#9 26.27 .../deasync@0.1.24/node_modules/deasync install$ node ./build.js
#9 26.37 .../deasync@0.1.24/node_modules/deasync install: gyp info it worked if it ends with ok
#9 26.37 .../deasync@0.1.24/node_modules/deasync install: gyp info using node-gyp@8.4.1
#9 26.37 .../deasync@0.1.24/node_modules/deasync install: gyp info using node@16.16.0 | linux | arm64
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python 
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python Python is not set from command line or npm configuration
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python Python is not set from environment variable PYTHON
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python checking if "python3" can be used
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python - "python3" is not in PATH or produced an error
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python checking if "python" can be used
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python - "python" is not in PATH or produced an error
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python 
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python **********************************************************
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python You need to install the latest version of Python.
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python Node-gyp should be able to find and use Python. If not,
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python you can try one of the following options:
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python - Use the switch --python="/path/to/pythonexecutable"
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python   (accepted by both node-gyp and npm)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python - Set the environment variable PYTHON
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python - Set the npm configuration variable python:
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python   npm config set python "/path/to/pythonexecutable"
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python For more information consult the documentation at:
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python https://github.com/nodejs/node-gyp#installation
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python **********************************************************
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! find Python 
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! configure error 
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack Error: Could not find any Python installation to use
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at PythonFinder.fail (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:330:47)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at PythonFinder.runChecks (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:159:21)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at PythonFinder.<anonymous> (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:202:16)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at PythonFinder.execFileCallback (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:294:16)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at exithandler (node:child_process:406:5)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at ChildProcess.errorhandler (node:child_process:418:5)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at ChildProcess.emit (node:events:527:28)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:289:12)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at onErrorNT (node:internal/child_process:478:16)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! System Linux 5.10.104-linuxkit
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! cwd /home/node/app/node_modules/.pnpm/deasync@0.1.24/node_modules/deasync
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! node -v v16.16.0
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! node-gyp -v v8.4.1
#9 26.41 .../deasync@0.1.24/node_modules/deasync install: gyp ERR! not ok 
#9 26.42 .../deasync@0.1.24/node_modules/deasync install: Build failed
```

</details>

File: src/data/api-service.ts
Comment (sarayourfriend) [reply src/data/api-service.ts:16]: It's already fixed in the API: https://github.com/WordPress/openverse-api/pull/801

This condition needs to stay, however, as the API isn't deployed with that fix yet.

File: .env.template
Comment (zackkrida) [inline .env.template:3]:
```suggestion
API_URL=http://127.17.0.1:8000/
API_CLIENT_ID=""
API_CLIENT_SECRET=""
``` 

Nit, maybe we uncomment these so there's one fewer step to edit the values once the user copies `.env.template` to `.env`?

File: TESTING_GUIDELINES.md
Comment (zackkrida) [inline TESTING_GUIDELINES.md:32]:
```suggestion
```

If we decide to remove the comments entirely (if not disregard).

File: TESTING_GUIDELINES.md
Comment (zackkrida) [inline TESTING_GUIDELINES.md:40]:
```suggestion
For the second, you'll need to follow [the instructions to set up an OAuth application](https://api.openverse.engineering/v1/#section/Register-and-Authenticate) and then fill in the `API_CLIENT_ID` and `API_CLIENT_SECRET` in the `.env` file copied from `.env.template`.
```

If we decide to remove the comments entirely (if not disregard).

File: TESTING_GUIDELINES.md
Comment (zackkrida) [inline TESTING_GUIDELINES.md:19]: This new section is great, very straightforward 👍

File: test/unit/specs/plugins/api-token.server.spec.ts
Comment (zackkrida) [inline test/unit/specs/plugins/api-token.server.spec.ts:162]: phenomenal documentation here, thanks

File: test/unit/test-utils/api-service-mock.ts
Comment (zackkrida) [inline test/unit/test-utils/api-service-mock.ts:40]:
```suggestion
 * With the particular way this mock creation function works,
```

File: .env.template
Comment (zackkrida) [reply .env.template:3]: Eh I'm not sold on this, and it's not important.

File: src/plugins/api-token.server.ts
Comment (zackkrida) [inline src/plugins/api-token.server.ts:133]:
```suggestion
  log('API client secrets found.')
```

For parity with the other log

File: src/plugins/api-token.server.ts
Comment (zackkrida) [inline src/plugins/api-token.server.ts:183]:
```suggestion
      log('injecting empty openverseApiToken into request context')
```

File: test/unit/.eslintrc.js
Comment (zackkrida) [inline test/unit/.eslintrc.js:13]: Nice!

File: test/unit/test-utils/api-service-mock.ts
Comment (zackkrida) [inline test/unit/test-utils/api-service-mock.ts:15]:
```suggestion
 * Creates an API Service that gets reused throughout the lifetime of
 * the current context in which it is called.
```

**Review by zackkrida — APPROVED** [review]
I just lost a long review comment I was writing to an accidental pagerefresh, but I spent a lot of time testing this and don't have nearly as much feedback as I'd hoped to be able to give. I applied @krysal's docker changes from #1609 and verified the correct behavior there, made sure that `pnpm dev` worked locally with API credentials that were invalid, empty, valid, and with no `.env` file. 

Awesome work.

**Review by obulat — COMMENTED** [review]
I couldn't run the Docker setup anyways, even with the changes from @krysal's PR (on that PR, Docker ran fine. Here, with the copy of those changes, it starts running and then stops for some reason. I would have to investigate more into why)
I have a really general question:
>This means that API calls during SSR will be authenticated, and not throttled, but API calls from the client (subsequent calls) will be throttled as an authenticated user because they won't have any API key set.
With the changes in this PR, will the client calls be throttled as _unauthenticated_ user? So, you would only be allowed _100 requests per day and 5 requests per hour rate limit_? 

I think I'm misunderstanding this PR, because if we only have 5 requests per hour, the frontend will be unusable. A single 'All content' search makes 2 requests (images & audio). So, 3 "All content" searches (or clicking on "Load more" or one or two filters selected) would overuse the limit...

**Comment by sarayourfriend** [comment]
> I think I'm misunderstanding this PR, because if we only have 5 requests per hour, the frontend will be unusable. A single 'All content' search makes 2 requests (images & audio). So, 3 "All content" searches (or clicking on "Load more" or one or two filters selected) would overuse the limit...

This is already the case, clients do not make authenticated or exempted requests. I do not know why throttling doesn't appear to be a problem. I tried to trigger the throttling against the production API recently and wasn't able to either. Keep in mind that clients individually making anonymous requests to the API do not add to the same throttle count as the server individually making requests to the API to fulfill SSR requests.

This PR doesn't change the client story at all, it's only changing the way the _server_ exempts itself from throttling.

**Comment by obulat** [comment]
Thank you for the explanation, @sarayourfriend !

> I do not know why throttling doesn't appear to be a problem. I tried to trigger the throttling against the production API recently and wasn't able to either.

This is weird...

**Comment by obulat** [comment]
I tried running this code (from Docker) against the newly-deployed dev api (https://api-dev.openverse.engineering), and it got an error response 401, 'www-authenticate': 'Bearer error="invalid_client"' when posting to '/v1/auth_tokens/token/' from SSR.

**Comment by obulat** [comment]
Here's what I'm doing to run front end in Docker with the local API:
1. Add `host.docker.internal` to the `allowed_hosts` in Django settings: 
```
ALLOWED_HOSTS = config("ALLOWED_HOSTS").split(",") + [
    gethostname(),
    gethostbyname(gethostname()),
    "host.docker.internal"
]
```
3. Create an `.env` file at the root of the front end repository and add `API_URL="http://localhost:8000/"` there, together with API_CLIENT_ID and API_CLIENT_SECRET.
4. docker build . --build-arg RELEASE=v0-test --build-arg API_URL=http://host.docker.internal:8000/ -t openverse-frontend:latest  (from the testing instructions,replacing the API_URL)
5. docker run --env-file .env -p 8443:8443/tcp openverse-frontend:latest

I don't get any results due to CORS error: `Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://host.docker.internal:8000/v1/images/?q=cat. (Reason: CORS request did not succeed). Status code: (null).` :(

File: src/data/api-service.ts
Comment (obulat) [reply src/data/api-service.ts]: @sarayourfriend , I think the API was deployed yesterday?

Comment (sarayourfriend) [reply src/data/api-service.ts]: Yup! Thanks :+1:

**Comment by sarayourfriend** [comment]
> I don't get any results due to CORS error: Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://host.docker.internal:8000/v1/images/?q=cat. (Reason: CORS request did not succeed). Status code: (null). :(

Is this on the client side or the server side? I've run into that problem on the client side a bunch when trying to run the frontend against a local API with everything in Docker. The problem is having to use a different URL on the client side than is used on the server side to hit the API, _I think_.

It's really far outside the scope of this change though. Please test this with `api-dev` as it's a much more realistic "true to prod" version. Everything, client and server, will be using the same URL to access the API.

**Comment by sarayourfriend** [comment]
I've just re-tested and confirmed that it works fine if I use `api-dev` for the API url rather than trying to do it locally.

If you need to test it locally then please don't test it with the docker container, do it by running `pnpm prod` with the `.env` file set up with local API url and client information.

**Comment by sarayourfriend** [comment]
Okay, so for me the 401 error ended up being because I missed the final character of my client ID in the env file. It was `xxxxxxxx` and it should have been `xxxxxxxxi`. This was just a bad copy paste job. Can you double check @obulat that your client information is correct in the .env file?

**Comment by zackkrida** [comment]
I can also confirm; not seeing any errors with the API_URL set to `https://api-dev.openverse.engineering/` and using my client and client secret that I registered and verified.

**Comment by obulat** [comment]
Woohoo 🎉 I can confirm that with `api-dev` this works well, and I see no errors!

**Review by obulat — APPROVED** [review]

**Pull request merged**
